core/trie2/triedb/pathdb/layertree.go

```diff
@@ -0,0 +1,198 @@
+package pathdb
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/core/trie2/trienode"
+	"github.com/NethermindEth/juno/core/trie2/trieutils"
+)
+
+// Represents a layer of the layer tree
+type layer interface {
+	// Returns the encoded node bytes for a given trie id, owner, path and isLeaf flag
+	node(id trieutils.TrieID, owner *felt.Felt, path *trieutils.Path, isLeaf bool) ([]byte, error)
+	// Updates the layer with a new root hash, state id and block number
+	update(root *felt.Felt, id, block uint64, nodes *nodeSet) *diffLayer
+	// Writes the journal to the given writer
+	journal(w io.Writer) error
+	// Returns the root hash of the layer
+	rootHash() *felt.Felt
+	// Returns the state id of the layer
+	stateID() uint64
+	// Returns the parent layer of the current layer
+	parentLayer() layer
+}
+
+// Represents a layer tree which contains multiple in-memory diff layers and a single disk layer
+type layerTree struct {
+	layers map[felt.Felt]layer
+	lock   sync.RWMutex
+}
+
+func newLayerTree(head layer) *layerTree {
+	tree := new(layerTree)
+	tree.reset(head)
+	return tree
+}
+
+// Returns the layer for a given root hash
+func (tree *layerTree) get(root *felt.Felt) layer {
+	tree.lock.RLock()
+	defer tree.lock.RUnlock()
+
+	return tree.layers[*root]
+}
+
+// Adds a new layer to the layer tree
+func (tree *layerTree) add(root, parentRoot *felt.Felt, block uint64, mergeClassNodes, mergeContractNodes *trienode.MergeNodeSet) error {
+	if root == parentRoot {
+		return errors.New("cannot have cycled layer")
```

Comment:
A better error here please, like "cyclic layers detected (root == parent root)" or something along this line